contributors/design-proposals/storage/container-storage-interface-cluster-registry.md

```diff
@@ -0,0 +1,158 @@
+
+
+## Background & Motivations
+
+Kubernetes supports the Container Storage Interface (CSI) to enable third party storage developers to deploy volume drivers exposing new storage systems in Kubernetes without having to touch the core Kubernetes code.
+Support for CSI was introduced as alpha in Kubernetes v1.9, and moved to beta in v1.10.
+See "CSI Volume Plugins in Kubernetes Design Doc" in the "Links" section below for details on how an arbitrary CSI volume driver is deployed and interacts with Kubernetes.
+The beta implementation of CSI has a number of limitations, including:
+* CSI drivers must be deployed with the provided CSI external-attacher sidecar container, even if they don’t implement ControllerPublishVolume.
+  * Meaning a CSI volume driver deployed on kubernetes has to deploy a CSI `external-attacher` container even if the volume driver doesn't require a "volume attach" operation.
+  * In this case, the CSI `external-attacher` container basically does a no-op in response to Kubernetes `VolumeAttachment` objects to allow Kubernetes to continue with the mounting process.
+* Users and cluster admins have no easy way to discover what CSI volume drivers are deployed on their Kubernetes cluster.
+
+The proposed "CSI cluster registration mechanism" should address these issues.
+
+### Links
+
+* [Container Storage Interface (CSI) Spec](https://github.com/container-storage-interface/spec/blob/master/spec.md)
+* [CSI Volume Plugins in Kubernetes Design Doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md)
+
+## Objective
+
+The objective of this document is to document all the requirements for enabling a cluster level registry for a CSI compliant volume drivers in Kubernetes.
+
+## Goals
+
+* Allow a CSI volume driver (that opts-in to new mechanism) to configure how Kubernetes should interact with it.
+* Improve discoverability by users/cluster admins of a CSI volume driver (that opts-in to new mechanism) deployed on the Kubernetes cluster.
+
+## Non-Goals
+
+* Define how CSI volume driver should be deployed on Kubernetes.
+* Require use of new CSI cluster registry mechanism for all CSI volume drivers deployed on Kubernetes (the mechanism will be optional/opt-in).
+
+## Design Overview
+
+A new custom resource will be automatically be installed on Kubernetes clusters.
+
+Upon deployment a driver must create a new custom resource object.
+
+## Design Details
+
+### `CSIDriver` Object
+
+#### Proposed API
+
+```go
+// CSIDriver captures information about a Container Storage Interface (CSI)
+// volume driver deployed on the cluster.
+//
+// CSIDriver objects are non-namespaced.
+type CSIDriver struct {
+    metav1.TypeMeta   `json:",inline"`
+
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+    metav1.ObjectMeta `json:"metadata"`
```

Comment:
Shall we require that ObjectMeta.Name is the CSI driver name? How else can we ensure that each CSI driver has just one CSIDriver instance?

Reply:
I don't think we should require it. We should have validation webhook to verify that you can't create two objects with the same `CSIDriverSpec.Driver` fields.

Reply:
This validation is IMO racy. There may be several API servers running in parallel.

1. API server A gets POST CSIDriver.name=X for a driver.
2. API server B gets POST CSIDriver.name=Y for the same driver at the same time.
3. Both call validation webhook.
4. The webhook gets two queries, possibly in parallel. Let's say the webhook synchronizes the requests and processes X fist.
5. The webhook can GET CSIDriver to check if it already exists, but in this case it does not return anything useful, X and Y were not created yet.
6. The webhook allows X to be created. It remembers that X exists in local cache.
7. The webhook processes Y, it can check in the API server if X exists. It does not exist yet, either the API server is slow or it was rejected by another webhook / validation.
8. The webhook can check internal cache if it allowed X. It was allowed, so it rejects Y.

And this is wrong, because X might have been rejected by API server (e.g. another webhook failed) and Y should have been allowed.